### PR TITLE
Always use -fmacro-prefix-map

### LINKED ISSUE #23195: Handling __FILE__ in code size tests in CircleCI
C++'s `__FILE__` macro expands to the file's path. Whether that's a relative path or an absolute one depends on what's given to the compiler. For example,
```console
$ clang++ ../../test.cpp
```
will expand `test.cpp`'s `__FILE__` into `../../test.cpp`, whereas
```console
$ clang++ ~/test.cpp
```
will expand it into `/whatever/absolute/path/test.cpp`.

In `system_lib.py`, we use both of them depending on the paths.
- We use relative paths when not using Ninja ([`build_objects`](https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/tools/system_libs.py#L483)) and `batch_inputs` is true, which is the default:
   https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/tools/system_libs.py#L536-L539
- We use absolute paths when not using Ninja ([`build_objects`](https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/tools/system_libs.py#L483)) and `batch_inputs` is false (#20929) 
- We use absolute paths when using Ninja ([`create_ninja_file`](https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/tools/system_libs.py#L168)). I don't think we can use relative paths in Ninja files.

In CircleCI, tests that depend on [`build-libs`](https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/.circleci/config.yml#L168-L200), which is a part of [`build-linux`](https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/.circleci/config.yml#L450-L488), which [uses Ninja](https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/.circleci/config.yml#L459), will be built with absolute paths when code includes `__FILE__`. This includes all core, other, and browser tests. Other tests (`test-mac-arm64`, `test-windows`, ...) don't use `build-libs` so they use `build_objects` and thus will include relative paths.

So, this will produce different builds for different CircleCI tests. `deterministic_paths` parameter, which uses `-ffile-prefix-map` in `system_lib.py` makes all relative paths the same string and all absolute paths the same string, but does not make an absolute path and a relative path the same. So even if we make CircleCI always set `deterministic_paths`, the problem remains:
https://github.com/emscripten-core/emscripten/blob/f615920e160c2d18f4cefa637c45745f0ffdff73/tools/system_libs.py#L495-L501

This problem was brought into attention because the new LLVM 19's libc++abi adds more usage of `__FILE__` and one of them ended up in `other.test_minimal_runtime_code_size_hello_embind`, causing the code sizes to be different between different CircleCI tests. We can make this usage an empty string in release mode (which Zig [did](https://github.com/ziglang/zig/commit/98a30acad6f28825cd93a9d478f6f2b84333bcb6)) but this does not fundamentally fix the problem that `__FILE__` can end up in other code size tests. Currently we seem to use `__FILE__` in several places in libraries:
```console
aheejin@aheejin:~/emscripten/system/lib$ grep __FILE__ * -R
compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h:      __sanitizer::CheckFailed(__FILE__, __LINE__, \
compiler-rt/lib/builtins/int_util.h:#define compilerrt_abort() __compilerrt_abort_impl(__FILE__, __LINE__, __func__)
libc/musl/include/assert.h:#define assert(x) ((void)((x) || (__assert_fail(#x, __FILE__, __LINE__, __func__),0)))
libcxx/include/__assert:       : _LIBCPP_ASSERTION_HANDLER(__FILE__ ":" _LIBCPP_TOSTRING(__LINE__) ": assertion " _LIBCPP_TOSTRING(            \
libcxx/src/verbose_abort.cpp:  __assert2(__FILE__, __LINE__, __func__, buffer);
libcxxabi/src/abort_message.h:        ::abort_message("%s:%d: %s", __FILE__, __LINE__, __msg);                                                       \
libcxxabi/src/abort_message.cpp:    __assert2(__FILE__, __LINE__, __func__, buffer);
mimalloc/include/mimalloc/types.h:#define mi_assert(expr)     ((expr) ? (void)0 : _mi_assert_fail(#expr,__FILE__,__LINE__,__func__))
wasmfs/support.h:  wasmfs::handle_unreachable(msg, __FILE__, __LINE__)
```

---

Not sure what is the best way to proceed:

1. Make `__FILE__` an empty string in release mode via adding `-D__FILE__=""` in `system_lib.py`.
2. Make `__FILE__` an empty string when the environment variable `CIRCLECI` is set, which is set in all CIrcleCI tests. But then we have to make sure to clear cache and set `CIRCLECI` when rebaselining code size tests on our local machine, which is a pain, like
   ```console
     $ ./emcc --clear-cache
     $ CIRCLECI=1 ./tools/maint/rebaseline_tests.py
   ```
3. ???

I personally prefer 1, which can be as simple as #23196.

### PR DESCRIPTION
**Pull request by aheejin**
When `deterministic_paths` is set, we are currently using `-ffile-prefix-map` to produce a deterministic prefix (`/emsdk/emscripten`) in data and debug info, which was recently done in #23222 and #23225. `-ffile-prefix-map` is an alias for both `-fmacro-prefix-map`, which affects `__FILE__` macro, and `-fdebug-prefix-map`, which affects the debug info section.

This lets the contents of `__FILE__` use the deterministic prefix `/emsdk/emscripten` unconditionally, i.e., even if `deterministic_paths` is not turned on. This is to produce reproducible builds and also make the code size tests' results consistent across platforms. The reason we still allow the real local path to be used in the debug info section when `deterministic_paths` is not set is to allow local developers to see their real paths.

This is basically implementing what's suggested in https://github.com/emscripten-core/emscripten/issues/23195#issuecomment-2549859260

This also turns `deterministic_paths` on for the Ninja path in embuilder for consistency with the non-Ninja path.

Fixes #23195.

### CODE REVIEW

File: tools/system_libs.py [hunk 1]
```diff
@@ -492,13 +497,15 @@ def build_objects(self, build_dir):
...
+    source_dir = utils.path_from_root()
+    relative_source_dir = os.path.relpath(source_dir, build_dir)
     if self.deterministic_paths:
-      source_dir = utils.path_from_root()
-      if batch_inputs:
-        relative_source_dir = os.path.relpath(source_dir, build_dir)
-        cflags += [f'-ffile-prefix-map={relative_source_dir}/=']
-      cflags += [f'-ffile-prefix-map={source_dir}=/emsdk/emscripten',
-                 '-fdebug-compilation-dir=/emsdk/emscripten']
+      cflags += [f'-ffile-prefix-map={relative_source_dir}={FAKE_EMSCRIPTEN_PATH}']
```
Comment (sbc100) [inline tools/system_libs.py:503]: Can you split this line out into its own PR?   Something like "system_libs.py: Use the deterministic path when building in batch mode".    This seems like a separate and uncontroversial bugfix.

Comment (aheejin) [reply tools/system_libs.py:503]: Done: #23222

Comment (sbc100) [reply tools/system_libs.py:503]: Great!

Once #23222 lands maybe the next step should just be to unconditionally enable `determinisic_paths`?  i.e. just remove the non-deterministic mode.    We can always add it back later if it turns out someone wants it... but I doubt they wil.

Comment (aheejin) [reply tools/system_libs.py:503]: I thought what we were going to do was to unconditionally use `-fmacro-prefix-map` but use `-ffile-prefix-map` (which also enables `-fdebug-prefix-map`) only in `deterministic_paths`? This is basically what https://github.com/emscripten-core/emscripten/issues/23195#issuecomment-2549859260 suggested but you also said maybe even that was not necessary... But anyway this PR does not remove the undeterministic mode. I can do either.

File: tools/system_libs.py [hunk 2]
```diff
@@ -600,12 +600,15 @@ def get_cflags(self):
...
 
+    source_dir = utils.path_from_root()
+    relative_source_dir = os.path.relpath(source_dir, self.build_dir)
     if self.deterministic_paths:
-      source_dir = utils.path_from_root()
-      relative_source_dir = os.path.relpath(source_dir, self.build_dir)
-      cflags += [f'-ffile-prefix-map={relative_source_dir}={DETERMINISITIC_PREFIX}']
       cflags += [f'-ffile-prefix-map={source_dir}={DETERMINISITIC_PREFIX}',
+                 f'-ffile-prefix-map={relative_source_dir}={DETERMINISITIC_PREFIX}',
                  f'-fdebug-compilation-dir={DETERMINISITIC_PREFIX}']
```
Comment (aheejin) [inline tools/system_libs.py:608]: `if deterministic_paths` part hasn't changed; I only merged two `cflags += ...` into one and changed the order of the flags.

File: tools/system_libs.py (hunk 1 above)
Comment (aheejin) [reply tools/system_libs.py]: I rebased this PR on top of #23222 and #23225. This still uses `deterministic_paths`. Please let me know if you think it'd better go unconditionally set `deterministic_paths` and remove the variable, i.e., you don't need your real local paths in the debug info.

Comment (sbc100) [reply tools/system_libs.py]: I think just removing `deterministic_paths` and always doing it this way makes sense.  Sorry for the back on forth on this.

**Comment by aheejin** [comment]
Will close in favor of #23256.
